### PR TITLE
Clean up

### PR DESCRIPTION
**Pull request by endersstocker**
Why:

* It could be cleaner.

How:

* Update `mix.lock`.
* Use Yoda-style assertions in tests.
* Remove `README.md` from WebDelivery.
* Create missing `WebDelivery.ErrorView` module.
* Remove unnecessary plugs.
* Replace `$APPLICATION_ENVIRONMENT` with `$DEPLOY_TAGS`.
* Remove _Dependencies_ section from `README.md`.
* Clean up files.
* Disable faulty Credo check.

### CODE REVIEW

File: elixir_buildpack.config [hunk 1]
```diff
@@ -1,4 +1,4 @@
 erlang_version=19.2
 elixir_version=1.4.0
 always_rebuild=true
-config_vars_to_export=(APPLICATION_ENVIRONMENT HOST_NAME ROLLBAR_ACCESS_TOKEN SECRET_KEY_BASE)
+config_vars_to_export=(HOST_NAME ROLLBAR_ACCESS_TOKEN SECRET_KEY_BASE)
```
Comment (bbrock25) [inline elixir_buildpack.config]: do you need DEPLOY_TAG in this list?

File: README.md [hunk 2]
```diff
@@ -145,26 +144,32 @@ Our exception tracking tool of choice is [Rollbar](https://rollbar.com). The
...
+$ heroku config:set ROLLBAR_ACCESS_TOKEN=#{access token copied from rollbar} DEPLOY_TAG=#{optional deploy tag to identify specific production environment}
 ```
 
-Once the environment variables are in place and the rollbar project is setup,
+Once the environment variables are in place and the rollbar project is set up,
 rollbar notifications can be created with:
 
 ```bash
-$ heroku addons:create deployhooks:http --url="https://api.rollbar.com/api/1/deploy/?access_token=#{access token copied from rollbar}&environment=production"
+$ heroku addons:create deployhooks:http --url="https://api.rollbar.com/api/1/deploy/?access_token=#{access token copied from rollbar}&environment=prod
```
Comment (samcdavid) [inline README.md]: You are better off using the DEPLOY_TAG instead of prod when creating the deploy hook.

File: README.md [hunk 3]
```diff
@@ -145,26 +144,32 @@ Our exception tracking tool of choice is [Rollbar](https://rollbar.com). The
...
+this with the following command:
 
 ```bash
-$ heroku config:set APPLICATION_ENVIRONMENT=production ROLLBAR_ACCESS_TOKEN=#{access token copied from rollbar}
+$ heroku config:set ROLLBAR_ACCESS_TOKEN=#{access token copied from rollbar} DEPLOY_TAG=#{optional deploy tag to identify specific production environment}
 ```
 
-Once the environment variables are in place and the rollbar project is setup,
+Once the environment variables are in place and the rollbar project is set up,
 rollbar notifications can be created with:
```
Comment (samcdavid) [inline README.md]: We should probably make mention that this the command they would run if the didn't set the deploy tag variable.

File: apps/web_delivery/web/web.ex [hunk 4]
```diff
@@ -28,15 +26,18 @@ defmodule WebDelivery.Web do
...
   end
 
   def view do
     quote do
       use Phoenix.View, root: "web/templates"
 
-      # Import convenience functions from controllers
-      import Phoenix.Controller, only: [get_csrf_token: 0, get_flash: 2, view_module: 1]
+      import Phoenix.Controller,
+        only: [get_csrf_token: 0, get_flash: 2, view_module: 1]
```
Comment (samcdavid) [inline apps/web_delivery/web/web.ex]: I don't think `get_flash` is needed since we aren't doing HTML views.

File: apps/web_delivery/web/web.ex [hunk 5]
```diff
@@ -55,11 +56,4 @@ defmodule WebDelivery.Web do
       import WebDelivery.Gettext
     end
   end
-
-  @doc """
-  When used, dispatch to the appropriate controller/view/etc.
-  """
-  defmacro __using__(which) when is_atom(which) do
-    apply(__MODULE__, which, [])
-  end
```
Comment (samcdavid) [inline apps/web_delivery/web/web.ex]: Was this moved because the convention is to define `__using__` first?

File: apps/web_delivery/web/views/error_view.ex [hunk 6]
```diff
@@ -0,0 +1,9 @@
+defmodule WebDelivery.ErrorView do
+  use WebDelivery.Web, :view
+
+  def render("404.json", _assigns), do: "Not found"
+
+  def render("500.json", _assigns), do: "Internal server error"
```
Comment (samcdavid) [inline apps/web_delivery/web/views/error_view.ex]: Since the request is expecting json, are we allowed to just respond with just a string? Should this be a map instead?

File: apps/web_delivery/test/controllers/status_controller_test.exs [hunk 7]
```diff
@@ -3,6 +3,6 @@ defmodule WebDelivery.StatusControllerTest do
 
   test "GET /api", %{conn: conn} do
     conn = get conn, "/api"
-    assert json_response(conn, 200) == %{"status" => "OK"}
+    assert %{"status" => "OK"} === json_response(conn, 200)
```
Comment (samcdavid) [inline apps/web_delivery/test/controllers/status_controller_test.exs]: Why was this change necessary?

File: apps/web_delivery/mix.exs [hunk 8]
```diff
@@ -36,4 +31,11 @@ defmodule WebDelivery.Mixfile do
       {:rollbax, "~> 0.8.0"},
     ]
   end
+
+  defp elixirc_paths(:test), do: ["lib", "web", "test/support"]
+  defp elixirc_paths(_),     do: ["lib", "web"]
+
+  def application do
+    [mod: {WebDelivery.Application, []}, extra_applications: [:logger]]
+  end
```
Comment (samcdavid) [inline apps/web_delivery/mix.exs]: Were these moved below deps because that is the convention?

File: apps/web_delivery/lib/web_delivery/endpoint.ex [hunk 9]
```diff
@@ -3,30 +3,17 @@ defmodule WebDelivery.Endpoint do
...
   plug Plug.Logger
 
   plug Plug.Parsers,
+    json_decoder: Poison,
     parsers: [:urlencoded, :multipart, :json],
-    pass: ["*/*"],
-    json_decoder: Poison
+    pass: ["*/*"]
 
-  plug Plug.MethodOverride
```
Comment (samcdavid) [inline apps/web_delivery/lib/web_delivery/endpoint.ex]: Why was this removed?

File: apps/web_delivery/lib/web_delivery.ex [hunk 10]
```diff
@@ -1,18 +1,2 @@
...
-      name: WebDelivery.Supervisor
-  end
-
-  # Tell Phoenix to update the endpoint configuration
-  # whenever the application is updated.
-  def config_change(changed, _new, removed) do
-    WebDelivery.Endpoint.config_change(changed, removed)
-    :ok
-  end
 end
```
Comment (samcdavid) [inline apps/web_delivery/lib/web_delivery.ex]: If everything else is defined using the single line style, can that be done for modules as well? Like
```elixir
demodule WebDelivery, do:
```

File: apps/web_delivery/lib/web_delivery/application.ex [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+      strategy: :one_for_one,
+      name: WebDelivery.Supervisor
+  end
+
+  @spec config_change(keyword, keyword, [atom]) :: :ok
+  def config_change(changed, _new, removed) do
+    Endpoint.config_change(changed, removed)
+    :ok
+  end
+end
```
Comment (samcdavid) [inline apps/web_delivery/lib/web_delivery/application.ex]: Can you help me understand why this was pulled out of `apps/web_delivery/lib/web_delivery.ex`?

File: apps/web_delivery/config/config.exs [hunk 12]
```diff
@@ -1,22 +1,19 @@
...
 
-# Configures Elixir's Logger
 config :logger, :console,
   format: "$time $metadata[$level] $message\n",
   metadata: [:request_id]
 
 config :rollbax,
   access_token: System.get_env("ROLLBAR_ACCESS_TOKEN"),
-  environment: System.get_env("APPLICATION_ENVIRONMENT") || Mix.env
+  environment: System.get_env("DEPLOY_TAG") || Mix.env
```
Comment (samcdavid) [inline apps/web_delivery/config/config.exs]: We'll need to verify the `DEPLOY_TAG` is in the elixir buildpack if it is going to be referenced here.

File: README.md [hunk 13]
```diff
@@ -176,12 +181,3 @@ $ heroku drains:add https://logs.loggly.com/inputs/#{CUSTOMER_TOKEN}/tag/product
...
 from other projects.
-
-## Dependencies
-
-- Erlang (latest)
-- Elixir (latest)
-- Hex
-- Rebar
-- Phoenix
-- Postgresql
```
Comment (samcdavid) [inline README.md]: Earlier in the doc we specify that a list of dependencies is included but they have been taken out. We need to leave them in here or alter the above copy. My vote is to leave the list of dependencies so that people who puke when they hear docker will at least know what they have to install.

File: elixir_buildpack.config (hunk 1 above)
Comment (endersstocker) [inline elixir_buildpack.config]: Nope. That config makes env vars available at compile time. Though `$DEPLOY_TAG` is evaluated at compile time, it is only used at runtime.

File: README.md (hunk 3 above)
Comment (endersstocker) [inline README.md]: Which command?

File: apps/web_delivery/web/web.ex (hunk 4 above)
Comment (endersstocker) [inline apps/web_delivery/web/web.ex]: You’re right. Nor is `get_csrf_token/0`.

File: apps/web_delivery/web/web.ex (hunk 5 above)
Comment (endersstocker) [inline apps/web_delivery/web/web.ex]: Exactly.

File: README.md (hunk 3 above)
Comment (samcdavid) [inline README.md]: The one on line 158.

File: apps/web_delivery/web/views/error_view.ex (hunk 6 above)
Comment (endersstocker) [inline apps/web_delivery/web/views/error_view.ex]: That’s a good question. `JSON.parse` is okay with it. According to _RFC-4627_ it’s invalid. According to `ECMA-404` it’s valid. It might be safer to err on the stricter side and return `"{\"error\":\"Not found\"}"` and `"{\"error\":\"Internal server error\"}"`. What do you think?

Comment (samcdavid) [inline apps/web_delivery/web/views/error_view.ex]: I personally would prefer the safe side, but would it be better to use a sigil instead of escaping the double quotes?

File: apps/web_delivery/test/controllers/status_controller_test.exs (hunk 7 above)
Comment (endersstocker) [inline apps/web_delivery/test/controllers/status_controller_test.exs]: There are actually two changes there: (1.) from equality to match equality and (2.) yoda-style assertion. Yoda-style assertions are commonplace in Elixir because they offer the ability to easily switch between comparison and pattern-matching. This comes in _particularly_ handy for maps. For example if I have this assertion:
```elixir
assert %{foo: 1, bar: 2, baz: 3} === some_fun(times)
```

...and I decide to add some metadata to it, say by turning it into a struct or backing it with an Ecto schema, and that metadata is unrelated to my assertion, all I have to do is change match equality to match, like so:
```elixir
assert %{foo: 1, bar: 2, baz: 3} = some_fun(times)
```

In order to make tests read consistently I recommend _always_ using yoda-style assertions in tests.

***

P.S. The style convention for multiline structures still applies with yoda-style assertions. This is pulled from the Guardian code:
```elixir
assert %{
  deposit_type: "wire",
  check_id: nil,
  check_number: nil,
} = changeset.changes
```

File: apps/web_delivery/web/views/error_view.ex (hunk 6 above)
Comment (endersstocker) [inline apps/web_delivery/web/views/error_view.ex]: 👍 It’s a matter of preference. I went with the string sigil as you suggested since it is probably more readable.

File: README.md (hunk 3 above)
Comment (endersstocker) [inline README.md]: Since the presence of a `DEPLOY_TAG` will be the exception and not the rule, I think it’s clear enough as is. I don’t feel too strong about that though, so I could be swayed.

Comment (samcdavid) [inline README.md]: It just feels like if we are going to show adding in the optional env var that the first example should use the env var to keep the flow easier to follow. That or where the env var is mentioned, we add in a second command that adds the optional environment variable. That way if you didn't use it, you would use the first example command for each case and if you did use it then you would use the second example command from each case. It just feels like if the flow is consistent it will be easier to grok when others have to use it.

File: apps/web_delivery/mix.exs (hunk 8 above)
Comment (endersstocker) [inline apps/web_delivery/mix.exs]: It’s all relative. 😆 Actually, `deps/0` and `elixirc_paths/1` were moved directly under `project/0` since they are private functions only used by it.

> If a private function is used by only one public function, place it directly below that function. If a private function is used by multiple public functions, place it below the the last function that uses it.

It is the organizational equivalent of lexical scoping. It’s especially helpful when your syntax highlighting is set to differentiate public and private functions from one another, like so:
public/private differentiation

File: apps/web_delivery/lib/web_delivery/endpoint.ex (hunk 9 above)
Comment (endersstocker) [inline apps/web_delivery/lib/web_delivery/endpoint.ex]: We are not using **POST** HTTP methods via HTML forms to emulate **PUT**, **PATCH**, and **DELETE** methods.

File: apps/web_delivery/mix.exs (hunk 8 above)
Comment (samcdavid) [inline apps/web_delivery/mix.exs]: Ok, that makes sense. It is just going to take a little time to get used to. I'm used to putting all public functions first in alphabetical order and then all private functions last in alphabetical order.

Comment (endersstocker) [inline apps/web_delivery/mix.exs]: Many languages (e.g., Ruby) force you to group your public and private methods, or else go through some ceremony (e.g., `private :this_method, :that_method`) that doesn’t help in the readability department. It’s a holdover from the days when source files were printed and cut in half to document the class API.

File: apps/web_delivery/lib/web_delivery.ex (hunk 10 above)
Comment (endersstocker) [inline apps/web_delivery/lib/web_delivery.ex]: You technically _can_ do `defmodule WebDelivery, do: nil`, but I wouldn’t recommend it.

File: apps/web_delivery/lib/web_delivery/application.ex (hunk 11 above)
Comment (endersstocker) [inline apps/web_delivery/lib/web_delivery/application.ex]: It’s cleaner, in general, to have your application callbacks in a dedicated module, but the more I think about it I doubt that any functions will be defined directly under the `WebDelivery` namespace. I’m actually going to merge the callbacks back in.

File: apps/web_delivery/config/config.exs (hunk 12 above)
Comment (endersstocker) [inline apps/web_delivery/config/config.exs]: If it’s not there, it just returns `nil`. It’s evaluated, but not used, during compilation. By runtime, when `:environment` is used, `$DEPLOY_TAG` will be available.

File: README.md (hunk 13 above)
Comment (endersstocker) [inline README.md]: If we include dependencies in the README, we need to address these issues:
* The current list is a subset of a mix of the primary and transitive dependencies.
* Erlang is a dependency of Elixir.
* The Elixir dependency is implicit.
* The Erlang and Elixir versions are already specified in `.tool-versions`.
* Hex tasks are included with Mix, which is a part of Elixir.
* Rebar is packaged with Elixir.
* Phoenix is already in WebDelivery’s `mix.exs` deps.
* Postgresql is not a dependency.

File: README.md (hunk 3 above)
Comment (endersstocker) [inline README.md]: 👍 Updated
